docs-internal/message-body-api.md

```diff
@@ -0,0 +1,918 @@
+2. `read`: if a charset is needed, derive it from the inbound `Content-Type` header is set defines a charset, otherwise use the server default charset
+
+Inbound headers should be read-only so that they are not modified by the readers. 
+
+Let's define a common algorithm for all writers such they are usable for both client and server:
+
+1. `accept`: the requested type is supported by the writer AND (the outbound `Content-Type` is set and accepted OR the writer's default `Content-Type` is accepted)
+2. `write`:
+  - a: if the outbound `Content-Type` header is not set, find the accepted `Content-Type` using the writer's default `Content-Type` as a fallback (in case of wildcard matching). This may throw an exception if not accepted value is found as a selected reader **must** have an accepted value
+  - b: if the outbound `Content-Type` header is not set, set it to the accepted `Content-Type` value
+  - c: if a charset is needed, derive it from the accepted `Content-Type` value if it defines a charset, otherwise use the server default charset
+
+An "accepted" `Content-Type` header is a header that is matching an `Accept` header value. A server response has a corresponding request which may have an `Accept` header value. If there no `Accept` header value (e.g. the request did not provide one, or it is a client request), then the everything is accepted ; i.e '*/*'.
+
+The underlying outbound headers should be writeable, writers can set any arbitrary header.
+
+Here are the header related methods provided by `MessageBodyReaderContext`:
+
+```java
+public ReadOnlyParameters headers() {
+  // uses ReadOnlyParameters to enforce read-only headers
+}
+
+public Optional<MediaType> contentType() {
+  // returns the Content-Type header value as a MediaType instance
+  // caches for optimization to avoid re-parsing every time
+}
+
+public boolean hasContentType(MediaType default) {
+  // designed to be used when implementing the accept method of a reader
+  // if the Content-Type header is set, return true if the default matches, false if not
+  // otherwise return true
+}
+
+@Override
+public Charset charset() throws IllegalStateException {
+  // designed to be used when implementing the read method of a reader
+  // derive the charset from the `Content-Type` value if set and the charset parameter is defined
+  // throws an IllegalStateException exception if the charset cannot be loaded
+  // otherwise use the server default
+}
+```
+
+Here are the header related methods provided by `MessageBodyWriterContext`:
+
+```java
+public Parameters headers() {
+  // uses Parameters to allow writes
+}
+
+public Optional<MediaType> contentType() {
+  // returns the Content-Type header value as a MediaType instance
+  // caches for optimization to avoid re-parsing every time
+}
+
+public void contentType(MediaType contentType) {
+  // sets the Content-Type header value if not already set
+}
+
+public void contentLength(long contentLength) {
```

Comment:
This method is not relevant in the case of a client request, add a note to state that the client writer context would use be a special implementation that would do a no-op.

Reply:
I do concur Sir.